### PR TITLE
BSIP 64: Optional HTLC Preimage Length and Add Hash160 Algorithm

### PR DESCRIPTION
**Pull request by jmjatlanta**
This BSIP will remove the requirement of a preimage length (making it optional), and adding an additional hash algorithm.

This PR fixes Issue https://github.com/bitshares/bsips/issues/163

### CODE REVIEW

**Comment by xeroc** [comment]
I like the solution that sets the size to 0 to disable length check!

+1

File: bsip-0064.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+This becomes even more important with atomic swaps. The shorter duration (a.k.a “inner”) contract should allow time to achieve its own irreversibility. And the longer duration (a.k.a. “outer”) contract must allow time for irreversibility of both.
+
+In addition, with an atomic swap both sides must consider the redemption time necessary. The creator of the inner contract must decide that should the redeemer redeem at the last moment, is there enough time to redeem the outer contract before the timelock expires.
+
+One final consideration of the timelock portion of the contract is the use of capital. Should the other party not accept, the funds in the contract are held until the timelock expires. Long expiration times could result in missed opportunity costs.
+
+Specifications
+=========
+Technically, the impelentation of these changes are small.
```
Comment (pmconrad) [inline bsip-0064.md]: Typo: impelentation

File: bsip-0064.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+**Preimage Size**: Permit that an HTLC can have the preimage size set to 0. This removes the size check during validation.
+
+**HASH160**: Include this additional hashing algorithm to the list of hashes available to be used.
+
```
Comment (pmconrad) [inline bsip-0064.md]: Please add that "Hash160" means "RIPEMD160(SHA256(x))".

File: bsip-0064.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+In addition, BIP199 includes the ability to use HASH160, which is not currently part of the BitShares implementation.
+
+Risks
+===============
+**Note** below is a discussion of HTLC risks in general, not risks inherent in implementing this BSIP.
+
+**An Oversized Preimage Attack** There are limits to the size of a preimage that can be used. That limit is set by the blockchain itself. But when dealing with atomic swaps across blockchains, an attacker could use a preimage that is too large for one side of the 2 sided transaction. That means that an HTLC can be created on both chains, but only redeemed on one chain, and not the other.
+
+This can be mitigated by specifying the preimage size as part of the contract. This feature is available on many bitcoin-based chains among others, but is not standard. Should the preimage size be used with an atomic swap, both sides of the transaction should include the preimage size in their contract.
+
```
Comment (pmconrad) [inline bsip-0064.md]: Perhaps add a note that even if the preimage size is not part of the explicit contract, implicit limitations may still exist (e. g. maximum transaction size of the blockchain). Another risk is that an explicit limitation may be incompatible with the implicit one, without anyone noticing.

File: bsip-0064.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+=========
+Technically, the impelentation of these changes are small.
+
+**Preimage Size**: Permit that an HTLC can have the preimage size set to 0. This removes the size check during validation.
+
+**HASH160**: Include this additional hashing algorithm to the list of hashes available to be used.
+
+Discussion
+===================
+https://github.com/bitshares/bsips/issues/163
```
Comment (pmconrad) [inline bsip-0064.md]: IMO this section should contain a brief summary and conclusion for important points that have been discussed. The link to the previous discussion is already in the header.

**Comment by jmjatlanta** [comment]
@pmconrad I believe I addressed your concerns. Please review and let me know if you have anything else. Thanks!

File: bsip-0064.md [hunk 5]
```diff
@@ -26,6 +26,8 @@ Risks
 
 This can be mitigated by specifying the preimage size as part of the contract. This feature is available on many bitcoin-based chains among others, but is not standard. Should the preimage size be used with an atomic swap, both sides of the transaction should include the preimage size in their contract.
 
+One item to keep in mind is that although a preimage size may not be explicitly specified, implicit limitations may exist. For insance, a blockchain may have limitations that make a transaction valid on one chain, but invalid on another (i.e. maximum transaction size).
```
Comment (pmconrad) [inline bsip-0064.md]: Typo: ins*t*ance

File: bsip-0064.md [hunk 6]
```diff
@@ -36,15 +38,19 @@ One final consideration of the timelock portion of the contract is the use of ca
...
 
 **Preimage Size**: Permit that an HTLC can have the preimage size set to 0. This removes the size check during validation.
 
-**HASH160**: Include this additional hashing algorithm to the list of hashes available to be used.
+**HASH160**: Include this additional hashing algorithm to the list of hashes available to be used. Note: A "HASH160" is simply RIPEMD160(SHA256(preimage)).
 
 Discussion
 ===================
-https://github.com/bitshares/bsips/issues/163
+While it is highly recommended that both halves of an atomic transaction provide explicit preimage lengths, the requirement as currently implemented in BitShares Core makes it incompatible with other "standardized" implementations. With the loostening of the preimage length requirement, atomic swaps that require such "standard" rules can be created on the BitShares blockchain.
```
Comment (pmconrad) [inline bsip-0064.md]: Typo: loos~~t~~ening

**Comment by pmconrad** [comment]
@nathanhourt have you seen https://github.com/bitshares/bitshares-fc/pull/143 ?

**Comment by pmconrad** [comment]
How about including this change here #202:
> Introducing this commonly used and known item for HTLC transaction as an optional feature is a very helpful step to satisfy the future of Cross Chain Swaps, parties involved within the Cross Chain Swaps will have to communicate their external addresses or some helpful references and info within the HTLC transaction upon creation.

**Comment by pmconrad** [comment]
Please specify that the memo field will be added as an extension to `htlc_create_operation`. This is sufficient to include it into the recipient's history, so there's no need to add it to the `htlc_object`.

**Pull request merged**